File: packages/shared/src/study/senses.ts
import { CardSense } from "../markdown/parser";

/**
 * Consecutive non-lapsed reviews after which a card counts as stable.
 *
 * This deliberately does NOT gate on `interval_days`. Under the fixed-cadence
 * scheduler Hard is always 1 day and Good always 2, so an interval only restates
 * which button was last pressed — an honest Good-presser would never cross a day
 * threshold. `repetitions` resets on a lapse and is otherwise independent of both
 * the rating and the scheduler, so it survives future scheduling changes.
 *
 * Spec 06 (depth ladders) must gate on this too rather than its own STABLE_DAYS.
 */
export const STABLE_REPS = 3;

export type WordPhase = "chain" | "split";

export interface SenseProgress {
  repetitions: number;
}

/**
 * A word stays chained until EVERY one of its senses is stable. Because a lapse
 * resets `repetitions`, graduation is reversible: failing one sense re-chains the
 * whole word so it gets mapped together again rather than drilled blind.
 */
export function wordPhase(senses: SenseProgress[]): WordPhase {
  if (senses.length === 0) return "chain";
  return senses.every((s) => s.repetitions >= STABLE_REPS) ? "split" : "chain";
}

export interface PromptCard {
  front: string;
  sense: CardSense | null;
}

/**
 * The prompt a client renders where the front would otherwise go. Chain hands the
 * sense over via the hint; split takes the scaffold away and asks with the context
 * sentence. Cards that are not senses always render their front, which is exactly
 * the behaviour every existing card already has.
 */
export function promptFor(card: PromptCard, phase: WordPhase): string {
  const sense = card.sense;
  if (!sense) return card.front;
  if (phase === "split") return sense.context ?? card.front;
  return sense.hint
    ? `${card.front} — ${sense.hint}?`
    : `${card.front} — meaning ${sense.index + 1} of ${sense.count}`;
}

File: android/app/src/main/java/com/flashmd/domain/senses/Senses.kt
package com.flashmd.domain.senses

import com.flashmd.data.parser.CardSense

/**
 * Mirror of packages/shared/src/study/senses.ts — keep in sync (Spec 10).
 *
 * Consecutive non-lapsed reviews after which a card counts as stable. This deliberately
 * does NOT gate on interval days: under the fixed-cadence scheduler Hard is always 1 day
 * and Good always 2, so an interval only restates which button was last pressed, and an
 * honest Good-presser would never cross a day threshold. `repetitions` resets on a lapse
 * and is otherwise independent of both the rating and the scheduler.
 */
const val STABLE_REPS = 3

enum class WordPhase { CHAIN, SPLIT }

/**
 * A word stays chained until EVERY sense is stable. Because `repetitions` resets on a
 * lapse, graduation is reversible: failing one sense re-chains the whole word.
 */
fun wordPhase(senseRepetitions: List<Int>): WordPhase =
    if (senseRepetitions.isNotEmpty() && senseRepetitions.all { it >= STABLE_REPS }) {
        WordPhase.SPLIT
    } else {
        WordPhase.CHAIN
    }

/**
 * The prompt rendered where the front would otherwise go. Chain hands the sense over via
 * the hint; split takes the scaffold away and asks with the context sentence. Cards that
 * are not senses always render their front — exactly today's behaviour.
 */
fun promptFor(front: String, sense: CardSense?, phase: WordPhase): String {
    if (sense == null) return front
    if (phase == WordPhase.SPLIT) return sense.context ?: front
    return if (sense.hint != null) {
        "$front — ${sense.hint}?"
    } else {
        "$front — meaning ${sense.index + 1} of ${sense.count}"
    }
}
